### PR TITLE
Shenli/global variable

### PR DESCRIPTION
**Pull request by shenli**
Get/Set global scope system variables from/to kv storage.

### CODE REVIEW

File: sessionctx/autocommit/autocommit.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// limitations under the License.
+
+package autocommit
+
+import (
+	"github.com/pingcap/tidb/context"
+)
+
+// Checker is the interface check if should autocommit in the context.
+type Checker interface {
```
Comment (ngaut) [inline sessionctx/autocommit/autocommit.go]: We should rename the name of the interface.

File: session.go [hunk 2]
```diff
@@ -258,6 +260,75 @@ func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (rset.Recor
...
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	log.Infof("GetGlobalSysVar over")
+	return value, nil
+}
+
+// SetGlobalSysVar implements RestrictedSQLExecutor.SetGlobalSysVar interface.
+func (s *session) SetGlobalSysVar(ctx context.Context, name string, value string) error {
+	sql := fmt.Sprintf(`UPDATE  %s.%s SET VARIABLE_VALUE="%s" WHERE VARIABLE_NAME="%s";`, mysql.SystemDB, mysql.GlobalVariablesTable, value, name)
```
Comment (zimulala) [inline session.go]: `name`  --> `strings.ToLower(name)` ?

File: sessionctx/autocommit/autocommit.go (hunk 1 above)
Comment (shenli) [inline sessionctx/autocommit/autocommit.go]: Any suggestion? It is already in autocommit package. So I don't want to add a "Autocommit" prefix.

**Comment by shenli** [comment]
PTAL

File: stmt/stmts/select.go [hunk 3]
```diff
@@ -149,7 +149,7 @@ func (s *SelectStmt) Plan(ctx context.Context) (plan.Plan, error) {
 		}
 	}
 	lock := s.Lock
-	if variable.ShouldAutocommit(ctx) {
+	if lock != coldef.SelectLockNone && ctx.(autocommit.Checker).ShouldAutocommit(ctx) {
```
Comment (siddontang) [inline stmt/stmts/select.go]: no need to check this type?

Comment (shenli) [inline stmt/stmts/select.go]: The only context we have is session. So we do not need to check it.
I use interface type asserting to solve dependence issues.

Comment (siddontang) [inline stmt/stmts/select.go]: I think you can add BindAutoCommitChecker(ctx, checker) and let session register it explicitly like other context mechanism does , and then you can use autocommit.ShouldAutoCommit(ctx) directly.

Comment (shenli) [inline stmt/stmts/select.go]: PTAL

File: session.go [hunk 4]
```diff
@@ -258,6 +261,75 @@ func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (rset.Recor
 	return rs, errors.Trace(err)
 }
 
+// GetGlobalSysVar implements RestrictedSQLExecutor.GetGlobalSysVar interface.
+func (s *session) GetGlobalSysVar(ctx context.Context, name string) (string, error) {
+	log.Infof("GetGlobalSysVar")
```
Comment (qiuyesuifeng) [inline session.go]: Remove this log.

File: session.go [hunk 5]
```diff
@@ -109,6 +111,7 @@ type session struct {
 	store   kv.Storage
 	sid     int64
 	history stmtHistory
+	initing bool
```
Comment (qiuyesuifeng) [inline session.go]: How about `inited`?

File: session.go [hunk 6]
```diff
@@ -258,6 +261,75 @@ func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (rset.Recor
...
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	defer rs.Close()
+	row, err := rs.Next()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	if row == nil {
+		return "", fmt.Errorf("Unknown sys var: %s", name)
```
Comment (qiuyesuifeng) [inline session.go]: errors.Errorf

File: session.go (hunk 5 above)
Comment (shenli) [inline session.go]: No, in bootstrap stage, autocommit checker will return false. For mysql.Global_variables table is not created. Access to this table will cause error. (AutocommitChecker should check autocommit variable, which is a system variable.)
So I use initing to indicate we are running bootstrap using this session.

**Comment by shenli** [comment]
PTAL

File: sessionctx/variable/sysvar.go [hunk 7]
```diff
@@ -571,3 +573,11 @@ const (
 	// CollationConnection is the name for collation_connection system variable.
 	CollationConnection = "collation_connection"
 )
+
+// GlobalSysVarAccessor is the interface for accessing global scope system variables.
+type GlobalSysVarAccessor interface {
+	// GetGlobalSysVar gets the global system variable value for name.
+	GetGlobalSysVar(ctx context.Context, name string) (string, error)
+	// SetGlobalSysVar set global system variable name to value.
```
Comment (qiuyesuifeng) [inline sessionctx/variable/sysvar.go]: sets

**Pull request merged**